Install via conda

mvdbeek opened the pull request:
This is @bwlang and my project for the gcc hackathon (also see https://github.com/bxlab/galaxy_hackathon/issues/39).
The actual place where we're doing the install still needs to move,
we should somehow show the requirement resolution to the admin in the UI
and the API could use a little reworking.

bwlang committed:
WIP re #39 - run dependency management during repository installation
working:
	- fetching valid tool list
	- identifying unique requirements

not working:
	File '/Users/langhorst/src/galaxy_hackathon/lib/galaxy/webapps/galaxy/controllers/admin_toolshed.py', line 1138 in prepare_for_install
  [view.manager_dependency(req) for req in uniq_reqs.values()]
TypeError: manager_dependency() takes exactly 1 argument (2 given)

mvdbeek committed:
Need to unnpack kwargs (ping @bwlang)

mvdbeek committed:
Move get_unique_requirements to shed_util_common

(ping @bwlang)

mvdbeek committed:
Iterate over all tools in a repository

mvdbeek committed:
Fix the parsing logic

mvdbeek committed:
Be a bit more defensive about parsing requirements

mvdbeek committed:
Fix indentation

bwlang committed:
passed a “manual install” argument to the dependency resolve
so that we can both respect a users wishes (independent of auto_install config)  and so that we can avoid attempting to 
create a conda_env in a nonexistent job directory

bwlang committed:
Probably better not to have a fake directory here…
other places check for job_directory=None

mvdbeek commented:
@bwlang might be a good idea to also include/review https://github.com/galaxyproject/galaxy/pull/2538/files

bwlang committed:
ask for tool dependency installation always.
I think a tool with no requirements is probably rare, 
and the new language should cover that situation.
I think it’s not worth the API call to fetch 
the requirements just to ask this question.

bwlang commented:
I think this is now ready for some feedback.
In my hands, it installs tools via conda.

Todo: give the user some feedback about whether the installation was successful.
not sure if we should try to do that on the tool installation status page, or on a dedicated requirements page. I see advantages to both approaches.

bwlang committed:
this view is not necessary for the current implementation

bwlang committed:
strangely… we need 
%if tool_dependencies_root_folder or missing_tool_dependencies_root_folder:

to avoid breaking the export repository code

bwlang committed:
resolving style issue
better comments

mvdbeek committed:
Add api route to fetch all requirements
and method that determines requirement status

This should be used in the UI to inform the admin of the dependency
resolution status. Also make galaxy package resolution more robust
to other files in the shed_tools directory (like condarc).

mvdbeek committed:
Continue dependency resolver  API routes and start work on displaying dependency status on a per repository view

bwlang committed:
Add first pass for displaying dependency status info

mvdbeek committed:
Add name and version to returned dependency

mvdbeek committed:
Hack in name and version into dictified representation of Dependency instance (For NullDependency we do want to show the missing requirements)

bwlang committed:
add status field with icons

bwlang commented:
@mvdbeek  what do you think...
ready to remove the WIP tag?

bgruening commented on templates/webapps/tool_shed/repository/common.mako:
```diff
@@ -1030,30 +1030,37 @@
 </%def>
 
-<%def name="render_resolver_dependency_items( resolver_dependencies )">
+<%def name="render_resolver_dependencies    ( resolver_dependencies )">
```
Can we remove the spaces?

bwlang committed:
whitespace fix

mvdbeek commented:
There are still a few things to sort out:
- [x] When the dependency is resolved with the versionless resolver, there should be a warning symbol (it's green right now. We should probably also differentiate between requested requriement version and resolved requirement version)
  - [x] There is the "CondaDepenency" typo in lib/, which is carried on into the UI
  - [ ] Toolshed packages are not shown as resolved (I think that may be in the nature of the toolshed resolver, not sure a "fix" would be simple)
  - [ ] We may be breaking the (display) of other resolver modules, we should at least test that those are working
  - [x] A set of tests would probably be a good idea.
  - [x] Return the resolution status through the API (https://github.com/galaxyproject/galaxy/pull/2554#discussion_r69380358)

We also discussed an overview page of all requirements ("Manage tool requirements" ?) with @jmchilton , so that the admin can see all requirements of all tools, the resolver status of these and a list of tools/repositories that depend on these requirements, but I think that would be another PR.

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -107,6 +108,35 @@ def __init__( self, config_filenames, tool_root_dir, app ):
+            for requirement in tool.requirements:
+                if requirement.type == 'package':
+                    reqs.append(json.dumps(requirement.to_dict()))
+        reqs = [json.loads(req) for req in list(set(reqs))]
```
No need to cast to list.

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -107,6 +108,35 @@ def __init__( self, config_filenames, tool_root_dir, app ):
+
+    def tools_requirements(self):
+        reqs = []
+        for desc, tool in self.app.toolbox.tools():
```
Maybe

```
s/self.app.toolbox.tools()/self.tools()/
```

is better?

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -107,6 +108,35 @@ def __init__( self, config_filenames, tool_root_dir, app ):
+        for desc, tool in self.app.toolbox.tools():
+            for requirement in tool.requirements:
+                if requirement.type == 'package':
+                    reqs.append(json.dumps(requirement.to_dict()))
```
You need to sort the dictionary keys here, otherwise 2 identical dicts may be serialized differently:

``` python
reqs.append(json.dumps(requirement.to_dict(), sort_keys=True))
```

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -107,6 +108,35 @@ def __init__( self, config_filenames, tool_root_dir, app ):
         )
+        self._view = views.DependencyResolversView(app)
+
+    def tools_requirements(self):
```
What about
s/tools_requirements/all_tools_requirements/
?

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -107,6 +108,35 @@ def __init__( self, config_filenames, tool_root_dir, app ):
+        reqs = []
+        for requirement in tool.requirements:
+            if requirement.type == 'package':
+                reqs.append(requirement.to_dict())
```
``` python
reqs = [req.to_dict() for req in tool.requirements if req.type == 'package']
```

mvdbeek committed:
Address @nsoranzo's review.

  - Skip unncessary list()
  - Sort keys when dumping requirements to json
  - Rename tools_requirements to all_tools_requirements
  - Use list comprehension where appropriate

nsoranzo commented on lib/galaxy/webapps/galaxy/api/tools.py:
```diff
@@ -110,6 +110,26 @@ def reload( self, trans, id, **kwd ):
     @web.require_admin
+    def list_requirements(self, trans, **kwds):
+        """
+        GET /api/tools/list_requirements
```
This is different from what is defined below in `lib/galaxy/webapps/galaxy/buildapp.py` , should be `/api/tools/all_requirements`

nsoranzo commented on lib/galaxy/webapps/galaxy/api/tools.py:
```diff
@@ -110,6 +110,26 @@ def reload( self, trans, id, **kwd ):
+    @web.require_admin
+    def tool_requirements_status(self, trans, id, **kwds):
+        """
+        GET /api/tools/{tool_id}/requirement_status
```
s/requirement_status/requirements/

mvdbeek replied:
Let me go through this, but my idea here was to return something like

``` python
 [{'name'='r-deseq2', 'resolver': 'CondaDependencyResolver', 'exact': True}, {'name'='r-edger', 'resolver': 'CondaDependencyResolver',  exact=False} ]
```

It doesn't do this yet (I'll add it to the things that still need to be done).
I could rename this to `requirements_status`, but I like the status in the name ...

mvdbeek replied:
In fact I should probably make all_requirements return the same structure ... huh, still some more work ahead :)

nsoranzo replied:
On the other hand, `requirements` is more general and can be extended to contain more information. Anyway, it should be synced with `lib/galaxy/webapps/galaxy/buildapp.py`

nsoranzo commented on lib/galaxy/webapps/galaxy/api/tools.py:
```diff
@@ -110,6 +110,26 @@ def reload( self, trans, id, **kwd ):
+    def tool_requirements_status(self, trans, id, **kwds):
+        """
+        GET /api/tools/{tool_id}/requirement_status
+        Return list of unique requirements for all tools.
```
This is copy/paste of above method documentation, should be updated.

nsoranzo commented on lib/galaxy/webapps/galaxy/buildapp.py:
```diff
@@ -252,11 +252,13 @@ def populate_api_routes( webapp, app ):
     webapp.mapper.connect( '/api/tools/{id:.+?}/diagnostics', action='diagnostics', controller="tools" )
     webapp.mapper.connect( '/api/tools/{id:.+?}/citations', action='citations', controller="tools" )
     webapp.mapper.connect( '/api/tools/{id:.+?}/download', action='download', controller="tools" )
+    webapp.mapper.connect( '/api/tools/{id:.+?}/requirement', action='tool_requirements_status', controller="tools")
```
s/requirement/requirements/

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -107,6 +108,32 @@ def __init__( self, config_filenames, tool_root_dir, app ):
+                if req == ireq['requirement']:
+                    req['status'] = "installed through %i" % ireq['index']
+                    result.append(req)
+        return result
```
I think these 2 methods (`tool_requirements()` and `tool_requirements_status()`) should probably be moved to class `Tool`.

mvdbeek committed:
Work on conda install revision

mvdbeek committed:
Fix dependency typo

mvdbeek committed:
Make api return more detailed requirements status, move tool_requirements and tool_requirements_status to Tool class

mvdbeek committed:
Use list comprehension to generate list of all requirements

mvdbeek committed:
Fix logic error in message

mvdbeek committed:
Display whether dependency is exactly resolved (check), inexactly resolved (warning) or not resolved (error)

mvdbeek committed:
Close docstring example list

mvdbeek committed:
Trigger requirement install for new install method

mvdbeek committed:
flake8 fixes

mvdbeek committed:
Add API endpoint to install (resolver) dependencies

mvdbeek committed:
Correct module description

mvdbeek committed:
Trigger dependency installation in InstallRepositoryManager

mvdbeek committed:
Add docstring to install_dependency

mvdbeek committed:
Revert manual_install mode (unncessary, better to use install_dependency)

mvdbeek committed:
Introduce installable_resolver property

mvdbeek committed:
Add install resolver dependency button

and add install_resolver_dependency to ToolShedRepositoriesController

mvdbeek committed:
Revert mako hack to always display install_tool_dependency button

 - Add resolver checkbox
 - Add possibility to disable resolvers if initialization failed

jmchilton commented:
This is amazing - what is left to do? Is there some chance to get some of this into 15.07?

mvdbeek commented:
Just need to fix the failing tests and rebase...  Other than that it's pretty complete by now.

mvdbeek committed:
Make conda verbose install check configurable

Move dependency installation from resolve() to install_dependency().

mvdbeek committed:
Only return requirements of type package (for now)

mvdbeek commented:
I think we're at a point where a review would be nice (ping @nsoranzo @jmchilton @bwlang @pvanheus @bgruening and others)!
There are currently some corner cases that are not how they should be (like tools with requirements but without any tool_dependencies.xml won't display the "install resolver dependencies" check box), but I think that could be done in a followup PR.
Also we changed a few things in how the dependency resolver status is shown,
it would be nice to know that we're not breaking anything related to the other dependency resolvers.

mvdbeek committed:
Use resolver.find_dep to determine dependency status

nsoranzo commented on lib/galaxy/tools/__init__.py:
```diff
@@ -107,6 +108,12 @@ def __init__( self, config_filenames, tool_root_dir, app ):
             tool_root_dir=tool_root_dir,
             app=app,
         )
+        self._view = views.DependencyResolversView(app)
```
This seems to be not used, can be removed?

mvdbeek replied:
Good catch, that's right.

nsoranzo commented on lib/galaxy/tools/deps/resolvers/__init__.py:
```diff
@@ -91,8 +92,12 @@ class NullDependency( Dependency ):
 
 
-INDETERMINATE_DEPENDENCY = NullDependency()
+INDETERMINATE_DEPENDENCY = NullDependency
```
Not sure if this change from a class instance to class alias is a good idea, there are still a number of places in `lib/` where the code is not updated to reflect this; also, if you need to change them all, then it would probably make sense to just use `NullDependency`.
@jmchilton: opinions?

mvdbeek replied:
Thanks, I found a few more instances where I had to update the code (b59b797).
The reason for the change is that I would like to use the resolver code to display whether a dependency can be resolved, but the NullDependency class is missing the dependencies' name and version.

I can also revert 0b24283, and [return to replicating the resolver logic](https://github.com/galaxyproject/galaxy/pull/2554/commits/0b242833959a5554d60488d6551da78c816020d4#diff-f623c3ca7acff7a0346735bed0aca07aL129) to determine whether a dependency is installed.
I kind of prefer how it is now, but either way is OK for me.

jmchilton replied:
Its fine to switch it like this - but I'd take @nsoranzo's advice and use NullDependency instead of INDETERMINATE_DEPENDENCY - the latter isn't a valid class name according to most Python naming conventions.

mvdbeek committed:
Update INDETERMINATE_DEPENDENCY use to NullDependency instance

mvdbeek committed:
Remove unused DependencyResolver view from ToolBox

mvdbeek committed:
Flake8 fix

mvdbeek committed:
Drop INDETERMINATE_DEPENDENCY alias

mvdbeek committed:
Fix NullDependency use

and add default arguments to NullDependency

mvdbeek committed:
Add api test for conda dependency resolution

nsoranzo commented:
@mvdbeek flake8 tests should be fixed now on `dev`, but I suspect you need to either rebase or merge `dev` to fix it in this PR.

mvdbeek commented on test/integration/test_resolvers.py:
```diff
@@ -0,0 +1,24 @@
+
+GNUPLOT = {u'version': u'4.6', u'type': u'package', u'name': u'gnuplot'}
+
+class CondaResolutionIntegrationTestCase(integration_util.IntegrationTestCase, ApiTestCase):
```
@jmchilton this test fails with 

```
ERROR: test suite for <class 'integration.test_resolvers.CondaResolutionIntegrationTestCase'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/nose/suite.py", line 209, in run
    self.setUp()
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/nose/suite.py", line 292, in setUp
    self.setupContext(ancestor)
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/nose/suite.py", line 315, in setupContext
    try_run(context, names)
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/nose/util.py", line 471, in try_run
    return func()
  File "/Users/marius/src/galaxy/test/base/integration_util.py", line 23, in setUpClass
    cls._test_driver.setup(config_object=cls)
  File "/Users/marius/src/galaxy/test/base/driver_util.py", line 641, in setup
    galaxy_config,
  File "/Users/marius/src/galaxy/test/base/driver_util.py", line 521, in launch_server
    static_enabled=True
  File "/Users/marius/src/galaxy/lib/galaxy/webapps/galaxy/buildapp.py", line 47, in app_factory
    return paste_app_factory( global_conf, **kwargs )
  File "/Users/marius/src/galaxy/lib/galaxy/webapps/galaxy/buildapp.py", line 84, in paste_app_factory
    webapp.add_ui_controllers( 'galaxy.webapps.galaxy.controllers', app )
  File "/Users/marius/src/galaxy/lib/galaxy/web/framework/webapp.py", line 115, in add_ui_controllers
    module = import_module( module_name )
  File "/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/Users/marius/src/galaxy/lib/galaxy/webapps/galaxy/controllers/admin_toolshed.py", line 32, in <module>
    from tool_shed.util import workflow_util
  File "/Users/marius/src/galaxy/lib/tool_shed/util/workflow_util.py", line 14, in <module>
    from galaxy.workflow.render import WorkflowCanvas
  File "/Users/marius/src/galaxy/lib/galaxy/workflow/render.py", line 1, in <module>
    import svgwrite
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/svgwrite/__init__.py", line 51, in <module>
    from svgwrite.drawing import Drawing
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/svgwrite/drawing.py", line 29, in <module>
    from svgwrite.container import SVG, Defs
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/svgwrite/container.py", line 27, in <module>
    from svgwrite.base import BaseElement
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/svgwrite/base.py", line 16, in <module>
    from svgwrite.params import Parameter
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/svgwrite/params.py", line 9, in <module>
    from svgwrite.validator2 import get_validator
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/svgwrite/validator2.py", line 9, in <module>
    from svgwrite.data import full11
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/svgwrite/data/full11.py", line 11, in <module>
    from svgwrite.data.typechecker import Full11TypeChecker as TypeChecker
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/svgwrite/data/typechecker.py", line 13, in <module>
    from svgwrite.data.svgparser import is_valid_transferlist, is_valid_pathdata, is_valid_animation_timing
  File "/Users/marius/src/galaxy/.venv/lib/python2.7/site-packages/svgwrite/data/svgparser.py", line 26, in <module>
    comma = Literal(',') * (0, 1)  # zero or one ','
TypeError: can't multiply sequence by non-int of type 'Literal'
```

Uncommenting the offending lines in svgwrite makes the tests pass ... any idea why this happens here, but not when running galaxy (or the regular API tests) ?

jmchilton replied:
Are you using nosetests from inside the venv or outside of it? This is really odd.

mvdbeek replied:
Or downgrading to svgwrite==1.1.3, which makes me think that svgwrite has a path problem, considering the [changelog](https://pypi.python.org/pypi/svgwrite)

mvdbeek replied:
from inside the venv ... I'll give outside a try.

jmchilton replied:
No - that is what you should be doing.

I think it breaks at this commit - https://bitbucket.org/mozman/svgwrite/commits/e661623be032d6060cf87f4f9df1bbf3b0d3011f when they make pyparsing and external dependency.

jmchilton replied:
`nosetests test/unit/workflows/test_render.py` works just fine. Must be a problem with import_module if using paste this way?

jmchilton replied:
So I thought hacking test_render might be a work around - modifying the entry point

```
import svgwrite

import os
from base import integration_util
from base.api import ApiTestCase
```

fixes the problem - modifying as

```
import os
from base import integration_util
from base.api import ApiTestCase
import svgwrite
```

Makes the error happen sooner.

Update:

Also fixes:

```
import os
from base import integration_util
import svgwrite
from base.api import ApiTestCase
```

mvdbeek replied:
Hmm, could that be a conflict with twill's copy of pyparsing?

jmchilton replied:
Update 2: Importing svgwrite before twill in test/base/twilltestcase.py is the exact point where this breaks down. Something in twill prevents svgwrite from working.

Like this:

```
# add extensions (twill/extensions) and the the wwwsearch & pyparsing  
```

Seems twill imports an older version of pyparsing - the older svgwrite didn't use pyparsing so this wasn't a problem.

mvdbeek replied:
Thanks a lot @jmchilton, importing svgwrite after `from base import integration_util` works.
That's good enough for a workaround, right?

mvdbeek committed:
Add more API tests for resolver dependencies

mvdbeek committed:
Fix syntax, return None if dependency not found

mvdbeek committed:
Revert svgwrite import, fixed in #2656

nsoranzo commented on test/integration/test_resolvers.py:
```diff
@@ -0,0 +1,86 @@
+    @classmethod
+    def handle_galaxy_config_kwds(cls, config):
+        cls.conda_tmp_prefix = mkdtemp()
+        cls.conda_tmp_prefix = '/tmp'
```
I guess you want only one of these 2 lines?

mvdbeek replied:
Oops, yes, you're right.

mvdbeek committed:
Remove reference to /tmp

jmchilton commented:
Awesome sauce - is this ready to go?

bgruening commented:
I did not had enough time to test everything, but the initial testing left a great impression.
@bwlang @mvdbeek thanks a lot!!!!

nsoranzo commented on lib/galaxy/tools/deps/__init__.py:
```diff
@@ -90,7 +90,7 @@ def dependency_shell_commands( self, requirements, **kwds ):
         for requirement in requirements:
             log.debug( "Building dependency shell command for dependency '%s'", requirement.name )
-            dependency = INDETERMINATE_DEPENDENCY
+            dependency = NullDependency(version=None, name=requirement.name)
```
`version=requirement.version`

nsoranzo commented on lib/galaxy/tools/deps/__init__.py:
```diff
@@ -115,10 +115,10 @@ def find_dep( self, name, version=None, type='package', **kwds ):
             if require_exact and not dependency.exact:
-                dependency = INDETERMINATE_DEPENDENCY
-            if dependency != INDETERMINATE_DEPENDENCY:
+                dependency = NullDependency(version=version, name=name)
```
Just `continue` ?

mvdbeek committed:
Use continue instead of unused NullDependency

in Dependency_Manager.find_dep(), and requirement.version instead of
None in Dependency_Manager.find_dep.dependency_shell_commands(). (thanks
@nsoranzo)

nsoranzo commented on lib/galaxy/tools/deps/resolvers/conda.py:
```diff
@@ -89,53 +88,57 @@ def get_option(name):
+                return NullDependency(version=version, name=name)
+
+        if not is_installed and self.auto_install:
+            is_installed = self.install_dependency(name, version, type)
```
`name=name, version=version, type=type`

nsoranzo commented on lib/galaxy/tools/deps/resolvers/conda.py:
```diff
@@ -166,34 +170,67 @@ def list_dependencies(self):
+            )
+            if not is_installed:
+                log.debug("Remove failed conda install of {}, version '{}'".format(name, version))
+                cleanup_failed_install(conda_target, conda_context=self.conda_context)
```
```
        if return_code != 0:
            is_installed = False
        else:
            # Recheck if installed
            is_installed = is_conda_target_installed(
                conda_target, conda_context=self.conda_context, verbose_install_check=self.verbose_install_check
            )
        if not is_installed:
            log.debug("Remove failed conda install of {}, version '{}'".format(name, version))
            cleanup_failed_install(conda_target, conda_context=self.conda_context)
```

nsoranzo commented on lib/galaxy/tools/deps/resolvers/galaxy_packages.py:
```diff
@@ -49,15 +49,15 @@ def _find_dep_default( self, name, type='package', exact=True, **kwds ):
             real_version = basename( real_path )
             return self._galaxy_package_dep(real_path, real_version, exact)
```
```
return self._galaxy_package_dep(real_path, real_version, name, exact)
```

mvdbeek replied:
good catch!

nsoranzo commented on lib/galaxy/tools/deps/resolvers/tool_shed_packages.py:
```diff
@@ -21,7 +21,7 @@ def _find_dep_versioned( self, name, version, type='package', **kwds ):
             path = self._get_package_installed_dependency_path( installed_tool_dependency, name, version )
             return self._galaxy_package_dep(path, version, True)
```
```
return self._galaxy_package_dep(path, version, name, True)
```

nsoranzo commented on lib/galaxy/tools/deps/resolvers/unlinked_tool_shed_packages.py:
```diff
@@ -57,21 +57,21 @@ def _find_dep_versioned( self, name, version, type='package', **kwds ):
             path = join( self.base_path, name, version )
             if exists(path):
                 # First try the way without owner/name/revision
                 package = self._galaxy_package_dep(path, version, True)
```
```
package = self._galaxy_package_dep(path, version, name, True)
```

nsoranzo commented on lib/galaxy/tools/deps/resolvers/unlinked_tool_shed_packages.py:
```diff
@@ -81,9 +81,9 @@ def _find_possible_depenencies(self, name, version, type):
                             for revision in listdir(package_path):
                                 revision_path = join(package_path, revision)
                                 package = self._galaxy_package_dep(revision_path, version, True)
```
```
package = self._galaxy_package_dep(revision_path, version, name, True)
```

nsoranzo commented on lib/galaxy/tools/deps/resolvers/unlinked_tool_shed_packages.py:
```diff
@@ -48,7 +48,7 @@ def _find_dep_versioned( self, name, version, type='package', **kwds ):
             possibles = self._find_possible_depenencies(name, version, type)
```
s/depenencies/dependencies/

mvdbeek committed:
Simplify conda install logic

nsoranzo commented on lib/galaxy/tools/deps/resolvers/unlinked_tool_shed_packages.py:
```diff
@@ -57,21 +57,21 @@ def _find_dep_versioned( self, name, version, type='package', **kwds ):
     # Should be extended as required
-    # Returns CandidateDepenency objects with data for preference picking
+    # Returns CandidateDependency objects with data for preference picking
     def _find_possible_depenencies(self, name, version, type):
```
s/depenencies/dependencies/

jmchilton commented:
I'd like to rework some of the user-facing GUI aspects of this:

Currently the wording is something like:
- [ ] Install resolver dependencies
- [ ] Install repository dependencies

I'd like it to be something more like:
- [ ] When available, install externally managed dependencies (e.g. conda) **_New**_
- [ ] When available, install tool shed managed dependencies
- Is this reasonable?
- Should we consider this beta? If so should we move the conda stuff after the shed stuff? The argument for it being considered beta is that the logging of the install process I think is a bit worse than that of the shed dependencies right?
- I want to get the conda checkbox to show up even if there is no tool_dependencies.xml file available. I'm going to take a crack at this.

martenson commented:
The UI does seem a bit confusing, mostly because it has been confusing before. 

Is there a reasonable scenario where you want to get both Conda and TS dependency installations? I feel like a radiobutton would be a better UI choice here. I also agree with @jmchilton that we should reword the text descriptions (like the 'un-check to skip' language)

> Should we consider this beta? If so should we move the conda stuff after the shed stuff?

Yes. Yes.

jmchilton commented:
@martenson I don't think the admin has a way of really knowing which they want - and for a large suite the answer is probably going to be both for a while.

I've been playing around with the biom stuff - where there is no tool_dependencies.xml and I do think with a small patch I have - things are more clear than they have even been.

jmchilton committed:
Show conda install option even when no tool_dependencies.xml is available.

Also reword different install options, I hope this is more clear - I do think it exposes les of the internal language to the end user.

martenson commented:
Also this builds on the older UI for tool installations. Would be nice to have this merged into the currently Beta API installation UI - https://github.com/galaxyproject/galaxy/pull/1392

The schedule of such is a bit tricky, I guess we continue with this PR as is and hope to port the changes to the new UI in 16.10?

mvdbeek committed:
Address @nsoranzo's review

nsoranzo commented:
Thanks @mvdbeek for fixing all my nitpicking! I leave it to @martenson and @jmchilton for more review, testing and hopefully merging!

jmchilton committed:
Mark conda install beta and move after TS options.

jmchilton committed:
Merge remote-tracking branch 'bxlab/install_via_conda' into install_via_conda

jmchilton commented:
Opened a PR with some of the changes @martenson  and I mentioned here: https://github.com/bxlab/galaxy_hackathon/pull/56.

martenson committed:
Merge pull request #56 from jmchilton/install_via_conda

UI Tweaks

jmchilton commented:
I'm getting an itchy merge finger ... and other important requests before this merged? I think we can make clarifications after the merge as usability bug fixes.

mvdbeek commented:
Awesome, thanks everyone!

mvdbeek commented:
> Also this builds on the older UI for tool installations. Would be nice to have this merged into the currently Beta API installation UI - #1392

This should work for new and old, we'll just have to change the wording to that of bxlab#56
<img width="830" alt="screen shot 2016-07-27 at 23 08 59" src="https://cloud.githubusercontent.com/assets/6804901/17190482/9bd63208-544e-11e6-952e-6137e8e424d9.png">
